crates/ibc/src/applications/transfer/memo.rs

```diff
@@ -0,0 +1,47 @@
+use core::convert::Infallible;
+use core::fmt::{self, Display};
+use core::str::FromStr;
+
+use crate::prelude::*;
+
+#[cfg_attr(
+    feature = "parity-scale-codec",
+    derive(
+        parity_scale_codec::Encode,
+        parity_scale_codec::Decode,
+        scale_info::TypeInfo
+    )
+)]
+#[cfg_attr(
+    feature = "borsh",
+    derive(borsh::BorshSerialize, borsh::BorshDeserialize)
+)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Memo(String);
```

Comment:
We need to add the memo to the events as well:

+ `TransferEvent` ([here](https://github.com/cosmos/ibc-go/blob/295bab01f633633e0d6136ba9b974c0e248edbe5/modules/apps/transfer/keeper/msg_server.go#L52) in ibc-go)
+ `RecvEvent` ([here](https://github.com/cosmos/ibc-go/blob/295bab01f633633e0d6136ba9b974c0e248edbe5/modules/apps/transfer/ibc_module.go#L203) in ibc-go)
+ `AckEvent` ([here](https://github.com/cosmos/ibc-go/blob/295bab01f633633e0d6136ba9b974c0e248edbe5/modules/apps/transfer/ibc_module.go#L250) in ibc-go)
+ `TimeoutEvent` ([here](https://github.com/cosmos/ibc-go/blob/295bab01f633633e0d6136ba9b974c0e248edbe5/modules/apps/transfer/ibc_module.go#L297) in ibc-go)

Reply:
Would it be possible to split the `Memo` part into a new PR, where we also upgrade ibc-proto-rs to v0.28.0 (first version with the memo field), and no tendermint-rs upgrade?

Essentially I don't think we should depend on tendermint v0.30+ until tower-abci releases a version which supports it; otherwise, it forces our users to use an unreleased version of tower-abci (probably their branch `tendermint-30`). In the meantime though we should still be able to support the memo field in ics 20

Reply:
Awesome thanks @erwanor! In this case, @kevinji , I think what's left for this PR is:
+ downgrade to use tendermint-rs v0.30 (to be compatible with tower-abci v0.6.0)
+ downgrade to use ibc-proto v0.29 (highest version which depends on tendermint-proto v0.30)
+ Add the memo to the events as mentioned above

Let me know if you need any help with any of this